Add new Python and C++/CUDA Custom Op tutorials (#2888)
pytorch/tutorials: advanced_source/cpp_custom_ops.rst

Custom C++ and CUDA Operators
=============================

.. note::
   This tutorial is for PyTorch 2.4+ and the PyTorch nightlies.

PyTorch offers a large library of operators that work on Tensors (e.g. torch.add, torch.sum, etc).
However, you may wish to bring a new custom operator to PyTorch. This tutorial demonstrates the
blessed path to authoring a custom operator written in C++/CUDA.

For our tutorial, we’ll demonstrate how to author a fused multiply-add C++
and CUDA operator that composes with PyTorch subsystems. The semantics of
the operation are as follows:

.. code-block:: python

  def mymuladd(a: Tensor, b: Tensor, c: float):
      return a * b + c

You can find the end-to-end working example for this tutorial over at
https://github.com/pytorch/extension-cpp .

Build System
------------

> If you author custom C++/CUDA code, it needs to be compiled somehow.

Reviewer: ```suggestion
If you are developing custom C++/CUDA code, it must be compiled.
```

Note that if you’re interfacing with a Python library that already has bindings

> to precompiled C++/CUDA code, then you may actually want to write a Python custom operator

Reviewer: ```suggestion
to precompiled C++/CUDA code, you might consider writing a customa Python operator instead.
```

    at::Tensor result = torch::empty(a_contig.sizes(), a_contig.options());
    const float* a_ptr = a_contig.data_ptr<float>();
    const float* b_ptr = b_contig.data_ptr<float>();
    float* result_ptr = result.data_ptr<float>();
  
    int numel = a_contig.numel();
    muladd_kernel<<<(numel+255)/256, 256>>>(numel, a_ptr, b_ptr, c, result_ptr);
    return result;
  }
  
  TORCH_LIBRARY_IMPL(extension_cpp, CUDA, m) {
    m.impl("mymuladd", &mymuladd_cuda);
  }

How to add torch.compile support for an operator
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

To add torch.compile support for an operator, we must add a FakeTensor kernel (also
known as a “meta kernel” or “abstract impl”). FakeTensors are Tensors that have
metadata (i.e. shape, dtype, device) but no data: the FakeTensor kernel for an
operator specifies how to compute the metadata of output tensors given the metadata of input tensors.

We recommend that this be done from Python via the `torch.library.register_fake` API,
though it is possible to do this from C++ as well (see
`The Custom Operators Manual <https://pytorch.org/docs/main/notes/custom_operators.html>`_
for more details).

> .. code-block:: python

Author: I need to figure out what orders are permissable

> .. code-block:: python

Author: Updated.

  def _backward(ctx, grad):
      a, b = ctx.saved_tensors
      grad_a, grad_b = None, None
      if ctx.needs_input_grad[0]:
          grad_a = torch.ops.extension_cpp.mymul.default(grad, b)
      if ctx.needs_input_grad[1]:
          grad_b = torch.ops.extension_cpp.mymul.default(grad, a)
      return grad_a, grad_b, None
  
  
  def _setup_context(ctx, inputs, output):
      a, b, c = inputs
      saved_a, saved_b = None, None
      if ctx.needs_input_grad[0]:
          saved_b = b
      if ctx.needs_input_grad[1]:
          saved_a = a
      ctx.save_for_backward(saved_a, saved_b)
  
  
  # This adds training support for the operator. You must provide us
  # the backward formula for the operator and a `setup_context` function
  # to save values to be used in the backward.
  torch.library.register_autograd(
      "extension_cpp::mymuladd", _backward, setup_context=_setup_context)

How to test an operator
-----------------------
Use torch.library.opcheck to test that the custom op was registered correctly.
This does not test that the gradients are mathematically correct; please write
separate tests for that (either manual ones or torch.autograd.gradcheck).

> .. code-block:: python

Author: I wrote them at different times. I can make them the same if we want.

    at::Tensor b_contig = b.contiguous();
    const float* a_ptr = a_contig.data_ptr<float>();
    const float* b_ptr = b_contig.data_ptr<float>();
    float* result_ptr = out.data_ptr<float>();
    for (int64_t i = 0; i < out.numel(); i++) {
      result_ptr[i] = a_ptr[i] + b_ptr[i];
    }
  }

When defining the operator, we must specify that it mutates the out Tensor in the schema:

.. code-block:: cpp
	TORCH_LIBRARY(extension_cpp, m) {
		m.def("mymuladd(Tensor a, Tensor b, float c) -> Tensor");
		m.def("mymul(Tensor a, Tensor b) -> Tensor");
		// New!
		m.def("myadd_out(Tensor a, Tensor b, Tensor(a!) out) -> ()");
	}

	TORCH_LIBRARY_IMPL(extension_cpp, CPU, m) {
		m.impl("mymuladd", &mymuladd_cpu);
		m.impl("mymul", &mymul_cpu);
		// New!
		m.impl("myadd_out", &myadd_out_cpu);
	}

Please do not return any mutated Tensors as outputs of the operator; this will
run you into problems later down the line.

Conclusion
----------
In this tutorial, we went over the recommended approach to integrating Custom C++
and CUDA operators with PyTorch. The TORCH_LIBRARY/torch.library APIs are fairly
low-level; more detail about how to use them can be found over at

> `The Custom Operators Manual <https://pytorch.org/docs/main/notes/custom_operators.html>`_

Reviewer: ```suggestion
`The Custom Operators Manual <https://pytorch.org/docs/main/notes/custom_operators.html>`_.
```

Custom C++ and CUDA Operators
=============================

.. note::
   This tutorial is for PyTorch 2.4+ and the PyTorch nightlies.

PyTorch offers a large library of operators that work on Tensors (e.g. torch.add, torch.sum, etc).
However, you may wish to bring a new custom operator to PyTorch. This tutorial demonstrates the
blessed path to authoring a custom operator written in C++/CUDA.

For our tutorial, we’ll demonstrate how to author a fused multiply-add C++
and CUDA operator that composes with PyTorch subsystems. The semantics of
the operation are as follows:

.. code-block:: python

  def mymuladd(a: Tensor, b: Tensor, c: float):
      return a * b + c

> You can find the end-to-end working example for this tutorial over at
> https://github.com/pytorch/extension-cpp .

Reviewer: ```suggestion
You can find the end-to-end working example for this tutorial `here <https://github.com/pytorch/extension-cpp>`__.
```

> 

Reviewer: Can we add What you will learn  and prerequisites like [here](https://github.com/pytorch/tutorials/blob/main/beginner_source/template_tutorial.py#L9)?
Example:
What you will learn
* How to author and integrate custom operators written in C++ 
* How to compile custom C++/CUDA code and registering it within the PyTorch framework to make it accessible from Python.
* Advanced integration techniques

Prerequisites
* PyTorch 2.4 or later
* Read `torch.utils.cpp_extension <https://pytorch.org/docs/stable/cpp_extension.html>`_
* Basic understanding of C++ and CUDA programming

we’ll focus on the "ahead of time" flavor.

Using ``cpp_extension`` is as simple as writing the following ``setup.py``:

.. code-block:: python

  from setuptools import setup, Extension
  from torch.utils import cpp_extension

  setup(name="extension_cpp",
        ext_modules=[
            cpp_extension.CppExtension("extension_cpp", ["muladd.cpp"])],
        cmdclass={'build_ext': cpp_extension.BuildExtension})

If you need to compile CUDA code (for example, ``.cu`` files), then instead use
`torch.utils.cpp_extension.CUDAExtension <https://pytorch.org/docs/stable/cpp_extension.html#torch.utils.cpp_extension.CUDAExtension>`_.
Please see how
`extension-cpp <https://github.com/pytorch/extension-cpp>`_ for an example for
how this is set up.

Defining the custom op and adding backend implementations
---------------------------------------------------------
First, let’s write a C++ function that computes ``mymuladd``:

.. code-block:: cpp

   at::Tensor mymuladd_cpu(at::Tensor a, const at::Tensor& b, double c) {
     TORCH_CHECK(a.sizes() == b.sizes());
     TORCH_CHECK(a.dtype() == at::kFloat);
     TORCH_CHECK(b.dtype() == at::kFloat);

>      TORCH_INTERNAL_ASSERT(a.device().type() == at::DeviceType::CPU);

Reviewer 2: nit: internal assert in user kernel is weird?

  @torch.library.register_fake("extension_cpp::mymuladd")
  def _(a, b, c):
      torch._check(a.shape == b.shape)
      torch._check(a.dtype == torch.float)
      torch._check(b.dtype == torch.float)
      torch._check(a.device == b.device)
      return torch.empty_like(a)

How to set up hybrid Python/C++ registration
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
In this tutorial, we defined a custom operator in C++, added CPU/CUDA
implementations in C++, and added ``FakeTensor`` kernels and backward formulas
in Python. The order in which these registrations are loaded (or imported)
matters (importing in the wrong order will lead to an error).

To use the custom operator with hybrid Python/C++ registrations, we must
first load the C++ library that holds the custom operator definition
and then call the ``torch.library`` registration APIs. This can happen in one
of two ways:

1. If you're following this tutorial, importing the Python C extension module
   we created will load the C++ custom operator definitions.
2. If your C++ custom operator is located in a shared library object, you can
   also use ``torch.ops.load_library("/path/to/library.so")`` to load it.

    

> How to add training (autograd) support for an operator

Reviewer: ```suggestion
Adding training (autograd) support for an operator
```

-- plan to write separate tests for that, either manual ones or by using
``torch.autograd.gradcheck``.

.. code-block:: python

  def sample_inputs(device, *, requires_grad=False):
      def make_tensor(*size):
          return torch.randn(size, device=device, requires_grad=requires_grad)
  
      def make_nondiff_tensor(*size):
          return torch.randn(size, device=device, requires_grad=False)
  
      return [
          [make_tensor(3), make_tensor(3), 1],
          [make_tensor(20), make_tensor(20), 3.14],
          [make_tensor(20), make_nondiff_tensor(20), -123],
          [make_nondiff_tensor(2, 3), make_tensor(2, 3), -0.3],
      ]
  
  def reference_muladd(a, b, c):
      return a * b + c
  
  samples = sample_inputs(device, requires_grad=True)
  samples.extend(sample_inputs(device, requires_grad=False))
  for args in samples:
      # Correctness test
      result = torch.ops.extension_cpp.mymuladd(*args)
      expected = reference_muladd(*args)
      torch.testing.assert_close(result, expected)
  
      # Use opcheck to check for incorrect usage of operator registration APIs
      torch.library.opcheck(torch.ops.extension_cpp.mymuladd.default, args)

.. _mutable-ops:

> How to create mutable operators

Reviewer: ```suggestion
Creating mutable operators
```
